Split and other changes

YoungFaithful opened the pull request:
- ts-delta-t
- OptVariable for input and lookup
- multiple years of data
- Storage formulation simplified
- Documentation Improvements
- CEP Split-up

YoungFaithful commented on REQUIRE:
```diff
@@ -1,6 +1,7 @@
 julia 0.7-
 Reexport
+ClustForOpt
```
ClustForOpt as dependency

YoungFaithful commented on data/CA_1/costs.csv:
```diff
@@ -0,0 +1,73 @@
+tech,location,year,account,USD,CO2
```
costs.csv as one file with additional column

YoungFaithful replied:
new folder -> `/data/{region}/`

YoungFaithful replied:
git mv that, but it seems to be off anyhow...

YoungFaithful commented on docs/make.jl:
```diff
@@ -1,19 +1,18 @@
 using Documenter
 using Plots
-using ClustForOpt
+using CEP
```
CEP details

YoungFaithful commented on docs/src/index.md:
```diff
@@ -1,38 +1,39 @@
-![ClustForOpt logo](assets/clust_for_opt_text.svg)
+![CEP logo](assets/cep_text.svg)
```
CEP details

YoungFaithful commented on docs/src/opt_cep.md:
```diff
@@ -1,4 +1,4 @@
-# Capacity Expansion Problem
+# Optimization Capacity Expansion Problem
```
restructuring

holgerteichgraeber replied:
Maybe call : Optimization problem formulation    ?

YoungFaithful commented on examples/workflow_example_cep.jl:
```diff
@@ -1,46 +1,55 @@
+years=[2015] #2016 works for GER_1 and CA_1, GER_1 can also be used with 2006 to 2016 and, GER_18 is 2015 TX_1 is 2008
 # laod ts-data
-ts_input_data, = load_timeseries_data("CEP", state; K=365, T=24) #CEP
+ts_input_data = load_timeseries_data_provided(state;T=24, years=years) #CEP
```
ending _provided to distinguish between data_path and state

holgerteichgraeber replied:
Is that function in ClustForOpt.jl? I think I've seen it there. 
I'm not sure if I fully follow why this is called `provided`.

holgerteichgraeber replied:
Never mind, comment from yesterday, I think we fixed this in the meantime.

YoungFaithful commented on examples/workflow_example_cep.jl:
```diff
@@ -1,46 +1,55 @@
-# select solver
-solver=ClpSolver()
+# select optimizer
+optimizer=Clp.Optimizer
```
jump 19

YoungFaithful commented on examples/workflow_example_cep.jl:
```diff
@@ -1,46 +1,55 @@
+
 # Use the design variable results for the operational run
-operation_result = run_opt(ts_full_data.best_results,cep_data,design_result.opt_config,get_cep_design_variables(design_result);solver=solver,lost_el_load_cost=1e6,lost_CO2_emission_cost=700)
+operation_result = run_opt(ts_full_data.best_results,cep_data,design_result.opt_config,design_variables,optimizer;lost_el_load_cost=1e6,lost_CO2_emission_cost=700)
```
second line for clarification

YoungFaithful commented on examples/workflow_introduction.jl:
```diff
@@ -0,0 +1,129 @@
+# To use this package
```
workflow from class

holgerteichgraeber replied:
Should we put parts of this into ClustForOpt as well?

YoungFaithful replied:
Yeah, I did so. it's also called `examples/workflow_introduction.jl`

YoungFaithful commented on src/CEP.jl:
```diff
@@ -0,0 +1,39 @@
+ # Capacity Expansion Probelem formulation
+ #
+ #####################
+module CEP
```
CEP split

YoungFaithful commented on src/optim_problems/run_opt.jl:
```diff
@@ -1,27 +1,26 @@
   #Check the consistency of the data provided
   check_opt_data_cep(opt_data)
-  cep=setup_opt_cep_basic(ts_data, opt_data, opt_config, solver; k_ids=k_ids)
+  cep=setup_opt_cep_basic(ts_data, opt_data, opt_config, optimizer, opt_config["optimizer_config"])
```
optimizer config to hold all configurations

holgerteichgraeber replied:
What are examples that can/should go into opt_config? Potentially add to documentation.

YoungFaithful replied:
That's a good idea. It is optimizer specific, but I could explain that for Gurobi.

YoungFaithful commented on src/optim_problems/run_opt.jl:
```diff
@@ -1,27 +1,26 @@
     setup_opt_cep_lost_emission!(cep, ts_data, opt_data)
   end
-  if opt_config["storage_p"] && opt_config["storage_e"] && opt_config["seasonalstorage"]
+  if opt_config["storage_in"] && opt_config["storage_out"] && opt_config["storage_e"] && opt_config["seasonalstorage"]
```
storage has separate technologies for `in` and `out` `p` is deprecated

holgerteichgraeber replied:
So default is p, and then e signifies energy component? Is this in the documentation?

YoungFaithful commented on src/optim_problems/run_opt.jl:
```diff
@@ -59,23 +58,22 @@ what you can add to the opt_config:
 
 """
-     run_opt(ts_data::ClustData,opt_data::OptDataCEP,fixed_design_variables::Dict{String,OptVariable};solver::Any=CbcSolver(),descriptor::String="",co2_limit::Number=Inf,lost_el_load_cost::Number=Inf,lost_CO2_emission_cost::Number=Inf,existing_infrastructure::Bool=false,intrastorage::Bool=false)
+     run_opt(ts_data::ClustData,opt_data::OptDataCEP,optimizer::DataTyple;descriptor::String="",co2_limit::Number=Inf,lost_el_load_cost::Number=Inf,lost_CO2_emission_cost::Number=Inf,existing_infrastructure::Bool=false,limit_infrastructure::Bool=false,storage::String="none",transmission::Bool=false,print_flag::Bool=true,print_level::Int64=0)
```
storage as simple String, not split up Bool

holgerteichgraeber replied:
Sounds good.

YoungFaithful commented on src/utils/optvariable.jl:
```diff
@@ -0,0 +1,317 @@
+#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
```
Make `OptVariable` rockin :+1:

YoungFaithful commented on test/cep.jl:
```diff
@@ -1,35 +1,60 @@
-
-    # run optimization
-    @test round(run_opt(ts_clust_res_8760.best_results,cep_input_data_GER;solver=ClpSolver(),storage="simple").objective)==round(run_opt(ts_clust_res_24.best_results,cep_input_data_GER;solver=ClpSolver(),storage="seasonal",k_ids=ts_clust_res_24.best_ids).objective)
+@testset "CEP" begin
```
testing is highly improved:
- connected testset as tree
- each single data element is checked in a separate test -> very detailed feedback if error occurs

YoungFaithful commented on test/cep_exact_data.jl:
```diff
@@ -0,0 +1,30 @@
+using CEP
```
the source code to calculate the exact_scenario
in case we ever change some of the input data.

YoungFaithful commented on src/optim_problems/opt_cep.jl:
```diff
@@ -1,45 +1,66 @@
+
   set=Dict{String,Array}()
-  set["nodes"]=unique(opt_data.nodes[:nodes])
+  set["nodes"]=axes(nodes,"node")
```
OptVariables are fancy now:
`axes(OptVariable,String)` gives the corresponding set

YoungFaithful commented on src/optim_problems/opt_cep.jl:
```diff
@@ -1,45 +1,66 @@
+    if occursin("tech",k) && String(k[end-1:end])=="_e"
+      setdiff!(set["tech_power"],v)
+      set["tech_energy"]=[set["tech_energy"];v]
+    end
```
more specific sets for setup of
- GEN
- CAP

YoungFaithful commented on src/optim_problems/opt_cep.jl:
```diff
@@ -129,26 +151,31 @@ function setup_opt_cep_lost_emission!(cep::OptModelCEP,
-                                  opt_data::OptDataCEP;
-                                  fixed_design_variables::Dict{String,OptVariable}=Dict{String,OptVariable}())
+                                  opt_data::OptDataCEP,
+                                  fixed_design_variables::Dict{String,Any})
```
that's a little annoying. Using Dict{String,OptVariable} errors in JLD2, because the lookup has a weird Vararg{} Type...

YoungFaithful commented on src/optim_problems/opt_cep.jl:
```diff
@@ -161,45 +188,41 @@ function setup_opt_cep_generation_el!(cep::OptModelCEP,
-    push!(cep.info,"COST['var',impact,tech] = Σ_{t,k,node}GEN['el',t,k,node]⋅ ts_weights[k] ⋅ ts_deltas ⋅ var_costs[tech,impact] ∀ impact, tech_generation")
-    @constraint(cep.model, [impact=set["impact"], tech=set["tech_generation"]], cep.model[:COST]["var",impact,tech]==sum(cep.model[:GEN]["el",tech,t,k,node]*ts_weights[k]*ts_deltas*find_cost_in_df(var_costs,nodes,tech,node,impact) for node=set["nodes"], t=set["time_T"], k=set["time_K"]))
+    push!(cep.info,"COST['var',impact,tech] = Σ_{t,k,node}GEN['el',t,k,node]⋅ ts_weights[k] ⋅ ts_deltas[t,k]⋅ var_costs[tech,impact] ∀ impact, tech_generation")
+    @constraint(cep.model, [impact=set["impact"], tech=set["tech_generation"]], cep.model[:COST]["var",impact,tech]==sum(cep.model[:GEN]["el",tech,t,k,node]*ts_weights[k]*ts_deltas[t,k]*costs[tech,node,set["year"][1],"var",impact] for node=set["nodes"], t=set["time_T"], k=set["time_K"]))
```
indexing as a normal array

YoungFaithful commented on src/optim_problems/opt_cep.jl:
```diff
@@ -161,45 +188,41 @@ function setup_opt_cep_generation_el!(cep::OptModelCEP,
     # Calculate Fixed Costs
-    push!(cep.info,"COST['cap_fix',impact,tech] = Σ_{node}CAP[tech,'new',node] ⋅ cap_costs[tech,impact] ∀ impact, tech_generation")
-    @constraint(cep.model, [impact=set["impact"], tech=set["tech_generation"]], cep.model[:COST]["cap_fix",impact,tech]==sum(cep.model[:CAP][tech,"new",node] *(find_cost_in_df(cap_costs,nodes,tech,node,impact)+find_cost_in_df(fix_costs,nodes,tech,node,impact)) for node=set["nodes"]))
+    push!(cep.info,"COST['cap_fix',impact,tech] = Σ_{t,k}(ts_weights ⋅ ts_deltas[t,k])/8760h ⋅ Σ_{node}CAP[tech,'new',node] ⋅ cap_costs[tech,impact] ∀ impact, tech_generation")
```
if multiple years are run
Assumption: The years repeat in the same year (2016, 2016, 2016) for the cost data as they are only monte carlo

holgerteichgraeber replied:
What do you mean by monte carlo?

YoungFaithful replied:
The actual weather data of each year has the same statistical chance to occur in the modelled year

YoungFaithful commented on src/optim_problems/opt_cep.jl:
```diff
@@ -216,45 +239,49 @@ function setup_opt_cep_storage!(cep::OptModelCEP,
-    push!(cep.info,"COST['fix',impact,tech] = Σ_{node}CAP[tech,'new',node] ⋅ cap_costs[tech,impact] ∀ impact, tech_storage")
-    @constraint(cep.model, [tech=[set["tech_storage_p"];set["tech_storage_e"]], impact=set["impact"]], cep.model[:COST]["cap_fix",impact,tech]==sum(cep.model[:CAP][tech,"new",node]*(find_cost_in_df(cap_costs,nodes,tech,node,impact)+find_cost_in_df(fix_costs,nodes,tech,node,impact)) for node=set["nodes"]))
+    push!(cep.info,"COST['fix',impact,tech] = Σ_{t,k}(ts_weights ⋅ ts_deltas[t,k])/8760h ⋅ Σ_{node}CAP[tech,'new',node] ⋅ costs[tech,node,year,'cap_fix',impact] ∀ impact, tech_storage")
+    @constraint(cep.model, [tech=[set["tech_storage_in"];set["tech_storage_out"];set["tech_storage_e"]], impact=set["impact"]], cep.model[:COST]["cap_fix",impact,tech]==sum(ts_weights[k]*ts_deltas[t,k] for t=set["time_T"], k=set["time_K"])/8760* sum(cep.model[:CAP][tech,"new",node]*costs[tech,node,set["year"][1],"cap_fix",impact] for node=set["nodes"]))
```
storage_p split up in two techs and GEN is used for each of them

YoungFaithful commented on src/optim_problems/opt_cep.jl:
```diff
@@ -416,29 +428,29 @@ Add co2 emission constraint
-                            co2_limit::Number=Inf)
+                            opt_data::OptDataCEP;
+                            co2_limit::Number=Inf,
+                            lost_emission_cost::Dict{String,Number}=Dict{String,Number}("CO2"=>Inf))
```
for multiple impact categories

YoungFaithful commented:
Ready to be inspected :)

YoungFaithful commented on docs/make.jl:
```diff
@@ -1,19 +1,22 @@
-        "Load Data" => "load_data.md",
-        "Clustering" => "clust.md",
-        "Optimization" => ["opt.md", "opt_cep.md", "opt_cep_data.md"]
+        "Data" => ["preparing_clust_data.md", "preparing_opt_cep_data.md", "csv_structure.md", "README_GER_18.md", "README_GER_1.md", "README_CA_14.md", "README_CA_1.md", "README_TX_1.md"],
```
New Structure into
- Data
- Optimization

holgerteichgraeber replied:
Great

YoungFaithful commented on src/CEP_development.jl:
```diff
@@ -0,0 +1,21 @@
+ #
+ #####################
+#using ClustForOpt
+include(joinpath("..","..","ClustForOpt","src","ClustForOpt_development.jl"))
```
For "Co-development"

holgerteichgraeber replied:
Can you walk me through the workflow here in person? Thanks

YoungFaithful commented on src/optim_problems/run_opt.jl:
```diff
@@ -59,23 +58,22 @@ what you can add to the opt_config:
                     opt_config::Dict{String,Any},
-                    fixed_design_variables::Dict{String,OptVariable};
-                    solver::Any=CbcSolver(),
+                    fixed_design_variables::Dict{String,Any},
```
JLD2 doesn't allow a Dict{String,OptVariable} anymore after OptVariable improvement. Seems to be a bug in JLD2 - Hope that this can be reverted in the future to OptVariable again

holgerteichgraeber replied:
Is there an issue in JLD2 concerning this, or anybody else having this problem? Is this a general problem with structs within a Dict?

YoungFaithful replied:
It's a specific issue in JLD2 with Dicts containing elements with Vararg{}

YoungFaithful commented on src/optim_problems/run_opt.jl:
```diff
@@ -30,24 +29,24 @@ function run_opt(ts_data::ClustData,
       setup_opt_cep_limit_infrastructure!(cep, ts_data, opt_data)
   end
-  setup_opt_cep_objective!(cep, ts_data, opt_data, opt_config["lost_load_cost"], opt_config["lost_emission_cost"])
+  setup_opt_cep_objective!(cep, ts_data, opt_data; lost_load_cost=opt_config["lost_load_cost"], lost_emission_cost=opt_config["lost_emission_cost"])
```
specified to reduce mixing it up.

holgerteichgraeber replied:
Sounds good

YoungFaithful commented on src/optim_problems/run_opt.jl:
```diff
@@ -115,171 +114,13 @@ function run_opt(ts_data::ClustData,
   #Run the optimization problem
-  run_opt(ts_data, opt_data, opt_config; solver=solver, k_ids=k_ids)
+  run_opt(ts_data, opt_data, opt_config, optimizer)
 end # run_opt
```
battery is in ClustForOpt

YoungFaithful commented on src/utils/load_data.jl:
```diff
@@ -1,83 +1,313 @@
-- `DAM`: Day Ahead Market
-- `CEP`: Capacity Expansion Problem
-and regions:
+load_timeseries_data_provided(region::String="GER_1"; T::Int64=24, years::Array{Int64,1}=[2016], att::Array{String,1}=Array{String,1}())
```
`load_time_series_data_provided()` now calls `load_time_series_data(data_path)` from ClustForOpt

YoungFaithful commented on src/utils/load_data.jl:
```diff
@@ -1,83 +1,313 @@
-- `cap_costs`:     `tech x [USD in USD/MW_el, CO2 in kg-CO₂-eq./MW_el]` # Entire (NOT annulized) Costs per Investment in technology
-- `techs`:        `tech x [categ,sector,lifetime in years,effic in %,fuel]`
-- `lines`:       `lines x [node_start,node_end,reactance,resistance,power,voltage,circuits,length]`
+    load_cep_data_techs(data_path::String)
```
Entire new structure:
- each former table is a OptVariable now: that means that they have sets and can be indexed like Arrays
- each nodes, costs, techs and lines has it's own load-function: Within the function the necessary values are extracted and the enry specific struct is filled: the value at each OptVariable entry is specific to be either
- Number
- OptDataCEPNode
- OptDataCEPLine
- OptDataCEPTech

the single OptDataCEP... hold specific information that needs to be provided. The load functions shall be expanded in the future to handle, if certain values are not provided -> automatically assign 0 or something similar and warn user about certain missing values in a nice way...

holgerteichgraeber replied:
Awesome! Do you have examples of that in the workflow example?

YoungFaithful replied:
Well, it's called when loading_cep_data. The user just needs to prepare the .csv files, which is explained in detail (https://youngfaithful.github.io/CEP.jl/dev/csv_structure/)

holgerteichgraeber commented on src/optim_problems/run_opt.jl:
```diff
@@ -1,27 +1,26 @@
 """
-    run_opt(ts_data::ClustData,opt_data::OptDataCEP,opt_config::Dict{String,Any};solver::Any=CbcSolver())
+    run_opt(ts_data::ClustData,opt_data::OptDataCEP,opt_config::Dict{String,Any};optimizer::DataType)
 organizing the actual setup and run of the CEP-Problem
```
```suggestion
organizing the actual setup and run of the CEP-Problem
- `opt_config`: Optimizer specific configuration data
```

YoungFaithful commented:
I worked suggestions into the documentation based on your suggestion. Some comments to your questions
- Monte Carlo years: Each weather year has the same statistical chance to happen in the modelled year
- `storage_p`: Doesn't exist any more. It is now split into `storage_in` and `storage_out`. I added an explanation into (https://youngfaithful.github.io/CEP.jl/dev/csv_structure/) - as soon as it is built
- JLD2 bug: It's a very specific bug with Dictionaries having an element, which contains a `Vararg{}` within its Type (`OptCEPVariable{Float64....Vararg{}}`)

holgerteichgraeber commented:
Great additions to the documentation!
